Genericize ARM64EHash into a platform-generic BufferChecksum and enable on x86

Achierius opened the pull request:
#### 6370e29508fef59a9b47e5a10b463fd36bf1c18b
<pre>
Genericize ARM64EHash into a platform-generic BufferChecksum and enable on x86
<a href="https://bugs.webkit.org/show_bug.cgi?id=276443">https://bugs.webkit.org/show_bug.cgi?id=276443</a>
<a href="https://rdar.apple.com/131483380">rdar://131483380</a>

Reviewed by NOBODY (OOPS!).

On x86, we substitute in a simple arithmetic cipher for the ARM64E PAC signing
functionality; this is not cryptographically secure, but it is no worse
from a security perspective than what we had before on x86 (nothing),
and importantly it allows us to catch memory corruptions which occur
between code-emission and link-time, which previously would only be
caught on x86 at the time the code was executed.

* Source/JavaScriptCore/assembler/AssemblerBuffer.h:
(JSC::BufferChecksum::BufferChecksum): Deleted.
(JSC::BufferChecksum::~BufferChecksum): Deleted.
(JSC::BufferChecksum::allocatePinForCurrentThreadAndInitializeHash): Deleted.
(JSC::BufferChecksum::deallocatePinForCurrentThread): Deleted.
(JSC::BufferChecksum::update): Deleted.
(JSC::BufferChecksum::makeDiversifier): Deleted.
(JSC::BufferChecksum::nextValue): Deleted.
(JSC::BufferChecksum::pin): Deleted.
(JSC::BufferChecksum::currentHash): Deleted.
(JSC::BufferChecksum::setUpdatedHash): Deleted.
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::copyCompactAndLinkCode):
(JSC::LinkBuffer::allocate):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6370e29508fef59a9b47e5a10b463fd36bf1c18b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38272 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11435 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62576 "Hash 6370e295 for PR 30843 does not build (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9387 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9589 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/62576 "Hash 6370e295 for PR 30843 does not build (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6693 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50980 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/62576 "Hash 6370e295 for PR 30843 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8271 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8391 "Hash 6370e295 for PR 30843 does not build (failure)") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52036 "Found 5 new JSC binary failures: testair, testapi, testb3, testdfg, testmasm, Found 108265 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/ArrayBtreeBadCodeGen.js.default, ChakraCore.yaml/ChakraCore/test/Array/ArrayElementMissingValueSetToZero.js.default, ChakraCore.yaml/ChakraCore/test/Array/InlineArrayPopWithIntConstSrc.js.default, ChakraCore.yaml/ChakraCore/test/Array/LastUsedSegmentHasNULLElement.js.default, ChakraCore.yaml/ChakraCore/test/Array/SegmentMapFlagResetInJSArrayConstructor.js.default, ChakraCore.yaml/ChakraCore/test/Array/TryGrowHeadSegmentBug.js.default, ChakraCore.yaml/ChakraCore/test/Array/arr_bailout.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_apply.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_ctr.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_fastinit.js.default ... (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54482 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8551 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64276 "Hash 6370e295 for PR 30843 does not build (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58185 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8511 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/64276 "Hash 6370e295 for PR 30843 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2866 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51007 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/64276 "Hash 6370e295 for PR 30843 does not build (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2395 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79946 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34101 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13825 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35185 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36270 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34931 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/assembler/AssemblerBuffer.h:
```diff
@@ -232,6 +241,24 @@ namespace JSC {
 
+        ALWAYS_INLINE uint32_t update(uint32_t instruction, uint32_t byteIdx)
+        {
+            uint32_t instrIdx = byteIdx / sizeof(instruction);
```
Would this penalize ARM64 perf for the sake of x64 code?  Maybe the compiler will optimize it out, but can you check?  This is in a perf sensitive area of code.

Achierius replied:
The two places we call `update` were previously:

AssemblerBuffer.h:463:
```
uint32_t hash = m_hash.update(value, m_index / sizeof(IntegralType));
```

LinkBuffer.h:323:
```
unsigned index = (bitwise_cast<uint8_t*>(ptr) - inData) / 4;
uint32_t hash = verifyUncompactedHash.update(value, index);
```

In both cases we were doing a /4 op beforehand, so all I've done is shift that to happening inside the function.

Since this function is `ALWAYS_INLINE` the resulting assembly is the same before and after; e.g. for the LinkBuffer.h call-site:
```
JavaScriptCore`void JSC::LinkBuffer::copyCompactAndLinkCode<unsigned int>(JSC::MacroAssembler&, JSC::JITCompilationEffort)::'lambda'(unsigned int const*)::operator()(unsigned int const*) const:
    0x10b009404 <+0>:  pacibsp 
    0x10b009408 <+4>:  sub    sp, sp, #0x40
    0x10b00940c <+8>:  stp    x29, x30, [sp, #0x30]
    0x10b009410 <+12>: add    x29, sp, #0x30
    0x10b009414 <+16>: stur   x0, [x29, #-0x8]
    0x10b009418 <+20>: stur   x1, [x29, #-0x10]
    0x10b00941c <+24>: ldur   x8, [x29, #-0x8]
    0x10b009420 <+28>: str    x8, [sp, #0x8]
```

I'll still run an A/B test on the final changes before/after to be sure, of course.

MenloDorian commented on Source/JavaScriptCore/assembler/AssemblerBuffer.h:
```diff
@@ -460,7 +512,7 @@ namespace JSC {
 #if CPU(ARM64)
             static_assert(sizeof(value) == 4);
 #if ENABLE(JIT_CHECKSUM_ASSEMBLER_BUFFER)
```
I don't see `JIT_CHECKSUM_ASSEMBLER_BUFFER` enabled for x64.  How is it used?

Also, have you checked if the x64 code flow will actually take this path that uses the `BufferChecksum`?

Achierius replied:
Good catch: I accidentally removed the chunk which did before submitting. I had checked (i.e. compiled+run on x86 and set a breakpoint on the appropriate function) but it must have been with the modified revision.

Achierius replied:
Fixed, it's now enabled on all platforms:

```
#if !defined(ENABLE_JIT_CHECKSUM_ASSEMBLER_BUFFER)
 #define ENABLE_JIT_CHECKSUM_ASSEMBLER_BUFFER 1
 #endif
 ```

MenloDorian commented on Source/JavaScriptCore/assembler/AssemblerBuffer.h:
```diff
@@ -232,6 +241,24 @@ namespace JSC {
             deallocatePinForCurrentThread();
         }
 
+        ALWAYS_INLINE uint32_t update(uint32_t instruction, uint32_t byteIdx)
```
This function was previously private.  Why does it have to be public now?

Achierius replied:
It was public before -- this is the main interface AssemblerBuffer uses to talk to the BufferChecksum.

Achierius committed:
Genericize ARM64EHash into a platform-generic BufferChecksum and enable on x86
https://bugs.webkit.org/show_bug.cgi?id=276443
rdar://131483380

Reviewed by NOBODY (OOPS!).

On x86, we substitute in a simple arithmetic cipher for the ARM64E PAC signing
functionality; this is not cryptographically secure, but it is no worse
from a security perspective than what we had before on x86 (nothing),
and importantly it allows us to catch memory corruptions which occur
between code-emission and link-time, which previously would only be
caught on x86 at the time the code was executed.

* Source/JavaScriptCore/assembler/AssemblerBuffer.h:
(JSC::BufferChecksum::BufferChecksum): Deleted.
(JSC::BufferChecksum::~BufferChecksum): Deleted.
(JSC::BufferChecksum::allocatePinForCurrentThreadAndInitializeHash): Deleted.
(JSC::BufferChecksum::deallocatePinForCurrentThread): Deleted.
(JSC::BufferChecksum::update): Deleted.
(JSC::BufferChecksum::makeDiversifier): Deleted.
(JSC::BufferChecksum::nextValue): Deleted.
(JSC::BufferChecksum::pin): Deleted.
(JSC::BufferChecksum::currentHash): Deleted.
(JSC::BufferChecksum::setUpdatedHash): Deleted.
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::copyCompactAndLinkCode):
(JSC::LinkBuffer::allocate):
